python/lsst/daf/butler/datastores/chainedDatastore.py

```diff
@@ -934,3 +947,75 @@ def export_records(self, refs: Iterable[DatasetIdRef]) -> Mapping[str, Datastore
+
+        if transfer == "move":
+            raise TypeError("Can not export by moving files out of datastore.")
+
+        # Exporting from a chain has the potential for a dataset to be
+        # in one or more of the datastores in the chain. We only need one
+        # of them since we assume the datasets are the same in all (but
+        # the file format could be different of course since that is a
+        # per-datastore configuration).
+        # We also do not know whether any of the datastores in the chain
+        # support file export.
+
+        # Ensure we have an ordered sequence that is not an iterator or set.
+        if not isinstance(refs, Sequence):
+            refs = list(refs)
+
+        # If any of the datasets are missing entirely we need to raise early
+        # before we try to run the export. This can be a little messy but is
+        # better than exporting files from the first datastore and then finding
+        # that one is missing but is not in the second datastore either.
+        known = [datastore.knows_these(refs) for datastore in self.datastores]
+        combined_known: dict[DatasetRef, bool] = {}
+        for known_to_this in known:
+            for ref, knows_this in known_to_this.items():
+                # If we have not known it yet, store the newer value.
+                if not combined_known.get(ref):
+                    combined_known[ref] = knows_this
+        missing_count = 0
+        for knows_this in combined_known.values():
+            if not knows_this:
+                missing_count += 1
+        if missing_count:
+            raise FileNotFoundError(f"Not all datasets known to this datastore. Missing {missing_count}")
+
+        # To allow us to slot each result into the right place after
+        # asking each datastore, create a dict with the index.
+        ref_positions = {ref: i for i, ref in enumerate(refs)}
+
+        # Presize the final export list.
+        exported: list[FileDataset | None] = [None] * len(refs)
+
+        # The order of the returned dataset has to match the order of the
+        # given refs, even if they are all from different datastores.
+        for i, datastore in enumerate(self.datastores):
+            known_to_this = known[i]
+            filtered = [ref for ref, knows in known_to_this.items() if knows and ref in ref_positions]
+
+            try:
+                this_export = datastore.export(filtered, directory=directory, transfer=transfer)
+            except NotImplementedError:
+                # Try the next datastore.
+                continue
+
+            for ref, export in zip(filtered, this_export):
+                # Get the position and also delete it from the list.
+                exported[ref_positions.pop(ref)] = export
+
+        for i, dataset in enumerate(exported):
+            if dataset is None:
+                raise FileNotFoundError(f"Failed to export dataset {refs[i]}.")
```

Comment:
Would it make sense to check that all datasets were exported before yielding any of them?

Reply:
I was trying to avoid scanning through the list of refs an additional time but I can do that.

Reply:
Thinking about it a bit more, this check is mostly there for mypy so it knows the method can't yield `None`.  The missing check earlier in the method should be enough to know that every dataset is known to at least one datastore. I can add a note that this is for mypy.